src/format.rs

```diff
@@ -0,0 +1,888 @@
+        assert_eq!(format(&time1, "'%8::z'").unwrap(), "'+00:00:00'");
+        assert_eq!(format(&time2, "'%8::z'").unwrap(), "'+00:09:21'");
+        assert_eq!(format(&time3, "'%8::z'").unwrap(), "'+01:00:00'");
+
+        assert_eq!(format(&time0, "'%-10::z'").unwrap(), "'-000:00:00'");
+        assert_eq!(format(&time1, "'%-10::z'").unwrap(), "'+000:00:00'");
+        assert_eq!(format(&time2, "'%-10::z'").unwrap(), "'+000:09:21'");
+        assert_eq!(format(&time3, "'%-10::z'").unwrap(), "'+001:00:00'");
+
+        assert_eq!(format(&time0, "'%_10::z'").unwrap(), "'  +0:00:00'");
+        assert_eq!(format(&time1, "'%_10::z'").unwrap(), "'  +0:00:00'");
+        assert_eq!(format(&time2, "'%_10::z'").unwrap(), "'  +0:09:21'");
+        assert_eq!(format(&time3, "'%_10::z'").unwrap(), "'  +1:00:00'");
+
+        assert_eq!(format(&time0, "'%8:::z'").unwrap(), "'+0000000'");
+        assert_eq!(format(&time1, "'%8:::z'").unwrap(), "'+0000000'");
+        assert_eq!(format(&time2, "'%8:::z'").unwrap(), "'+00:09:21'");
+        assert_eq!(format(&time3, "'%8:::z'").unwrap(), "'+0000001'");
+
+        assert_eq!(format(&time0, "'%-10:::z'").unwrap(), "'-000000000'");
+        assert_eq!(format(&time1, "'%-10:::z'").unwrap(), "'+000000000'");
+        assert_eq!(format(&time2, "'%-10:::z'").unwrap(), "'+000:09:21'");
+        assert_eq!(format(&time3, "'%-10:::z'").unwrap(), "'+000000001'");
+
+        assert_eq!(format(&time0, "'%-_10:::z'").unwrap(), "'        -0'");
+        assert_eq!(format(&time1, "'%-_10:::z'").unwrap(), "'        +0'");
+        assert_eq!(format(&time2, "'%-_10:::z'").unwrap(), "'  +0:09:21'");
+        assert_eq!(format(&time3, "'%-_10:::z'").unwrap(), "'        +1'");
+
+        let time4 = Time::new(1, 1, 1, 1, 1, 1, 1, Offset::fixed(0).unwrap()).unwrap();
+        let time5 = Time::new(1, 1, 1, 1, 1, 1, 1, TimeZoneRef::utc().into()).unwrap();
+
+        assert_eq!(format(&time0, "'%10Z'").unwrap(), "'       UTC'");
+        assert_eq!(format(&time1, "'%10Z'").unwrap(), "'       UTC'");
+        assert_eq!(format(&time2, "'%10Z'").unwrap(), "'       LMT'");
+        assert_eq!(format(&time3, "'%-^#10Z'").unwrap(), "'cet'");
+        assert_eq!(format(&time4, "'%010Z'").unwrap(), "'00000+0000'");
+        assert_eq!(format(&time5, "'%010Z'").unwrap(), "''");
+
+        // TODO: other tests
+
+        // Time.new(2022).strftime("%A\u{30a}") == "Saturday\u{30a}"
+        // Time.new(2022).strftime("%\u{c5}") == "%\u{c5}"
+
+        // Time.new(4).strftime("%9%y") == "   %y"
+        // Time.new(4).strftime("%9%%y") == "   %04"
+    }
+
+    #[test]
+    fn test_format_too_large() {
+        let time_zone_ref = tzdb::tz_by_name("Europe/Paris").unwrap();
+        let time = Time::new(1, 1, 1, 1, 1, 1, 1, time_zone_ref.into()).unwrap();
+
+        assert!(matches!(
+            format(&time, "%100000000Y").unwrap_err(),
+            FormatError::IoError(err @ io::Error { .. }) if err.kind() == io::ErrorKind::WriteZero,
```

Comment:
Is it possible to transform this into something that is closer to `ERANGE`?

```
[3.1.2] > Time.now.strftime "%100000000Y"
(irb):10:in `strftime': Result too large - %100000000Y (Errno::ERANGE)
        from (irb):10:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
```

Reply:
Now that I know about https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.WriteZero, I think we can totally detect this case in `impl From<io::Error> for FormatError` and transform this case into a new `FormatError::RangeError` (name TBD).